File: packages/android/src/main/kotlin/com/napaxi/android/ConfigStore.kt
package com.napaxi.android

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

public data class NapaxiConfigProfile(
    val id: String,
    val name: String,
    val provider: String,
    val model: String,
    val baseUrl: String? = null,
    val systemPrompt: String = "",
    val maxTokens: Int = LlmConfig.DEFAULT_MAX_TOKENS,
    val maxToolIterations: Int = 50,
    val extraHeaders: String? = null,
    val userTimezone: String? = null,
    val allowedModels: List<Map<String, String>>? = null,
    val imageModel: String? = null,
    val imageAnalysisModel: String? = null,
    val videoModel: String? = null,
    val audioModel: String? = null,
    val contextEngine: ContextEngineConfig = ContextEngineConfig(),
    val shellSecurity: ShellSecurityConfig = ShellSecurityConfig(),
    val metadata: JSONObject = JSONObject(),
) {
    public fun toJsonObject(): JSONObject = JSONObject()
        .put("id", id)
        .put("name", name)
        .put("provider", provider)
        .put("base_url", baseUrl)
        .put("model", model)
        .put("max_tokens", maxTokens)
        .put("max_tool_iterations", maxToolIterations)
        .put("extra_headers", extraHeaders)
        .apply {
            if (systemPrompt.trim().isNotEmpty()) put("system_prompt", systemPrompt)
            userTimezone?.takeIf { it.trim().isNotEmpty() }?.let { put("user_timezone", it) }
            allowedModels?.let { models ->
                put(
                    "allowed_models",
                    JSONArray(
                        models.map { model ->
                            JSONObject().apply {
                                model.forEach { (key, value) -> put(key, value) }
                            }
                        },
                    ),
                )
            }
            imageModel?.let { put("image_model", it) }
            imageAnalysisModel?.let { put("image_analysis_model", it) }
            videoModel?.let { put("video_model", it) }
            audioModel?.let { put("audio_model", it) }
            put("context_engine", contextEngine.toJsonObject())
            put("shell_security", shellSecurity.toJsonObject())
            if (metadata.length() > 0) put("metadata", metadata)
        }

    public fun toJson(): String = toJsonObject().toString()

    public fun toMap(): Map<String, Any?> = toJsonObject().toPlainMap()

    public fun toConfig(apiKey: String): LlmConfig =
        LlmConfig(
            provider = provider,
            apiKey = apiKey,
            baseUrl = baseUrl,
            model = model,
            systemPrompt = systemPrompt,
            maxTokens = maxTokens,
            maxToolIterations = maxToolIterations,
            extraHeaders = extraHeaders,
            userTimezone = userTimezone,
            allowedModels = allowedModels,
            imageModel = imageModel,
            imageAnalysisModel = imageAnalysisModel,
            videoModel = videoModel,
            audioModel = audioModel,
            contextEngine = contextEngine,
            shellSecurity = shellSecurity,
        )

    public companion object {
        @JvmStatic
        public fun fromJson(rawJson: String): NapaxiConfigProfile =
            fromJsonObject(JSONObject(rawJson))

        @JvmStatic
        public fun fromJsonObject(obj: JSONObject): NapaxiConfigProfile =
            NapaxiConfigProfile(
                id = obj.optString("id"),
                name = obj.optString("name"),
                provider = obj.optString("provider"),
                baseUrl = obj.optNullableString("base_url"),
                model = obj.optString("model"),
                systemPrompt = obj.optString("system_prompt"),
                maxTokens = obj.optInt("max_tokens", LlmConfig.DEFAULT_MAX_TOKENS),
                maxToolIterations = obj.optInt("max_tool_iterations", 50),
                extraHeaders = obj.optNullableString("extra_headers"),
                userTimezone = obj.optNullableString("user_timezone")
                    ?: obj.optNullableString("userTimeZone")
                    ?: obj.optNullableString("timeZoneId"),
                allowedModels = obj.optJSONArray("allowed_models")?.toStringMapList(),
                imageModel = obj.optNullableString("image_model"),
                imageAnalysisModel = obj.optNullableString("image_analysis_model"),
                videoModel = obj.optNullableString("video_model"),
                audioModel = obj.optNullableString("audio_model"),
                contextEngine = obj.optJSONObject("context_engine")?.let(ContextEngineConfig::fromJsonObject)
                    ?: ContextEngineConfig(),
                shellSecurity = obj.optJSONObject("shell_security")?.let(ShellSecurityConfig::fromJsonObject)
                    ?: ShellSecurityConfig(),
                metadata = obj.optJSONObject("metadata") ?: JSONObject(),
            )

        @JvmStatic
        public fun fromMap(map: Map<String, *>): NapaxiConfigProfile =
            fromJsonObject(JSONObject(map))
    }
}

public data class NapaxiConfigSelection(
    val selectedProfileId: String? = null,
    val selectedProfileIdByCapability: Map<String, String> = emptyMap(),
    val systemPrompt: String = "",
    val maxToolIterations: Int = 50,
) {
    public fun toJsonObject(): JSONObject = JSONObject()
        .put("selected_profile_id", selectedProfileId)
        .put("selected_profile_id_by_capability", JSONObject(selectedProfileIdByCapability))
        .put("system_prompt", systemPrompt)
        .put("max_tool_iterations", maxToolIterations)

    public fun toJson(): String = toJsonObject().toString()

    public fun toMap(): Map<String, Any?> = toJsonObject().toPlainMap()

    public companion object {
        @JvmStatic
        public fun fromJson(rawJson: String): NapaxiConfigSelection =
            fromJsonObject(JSONObject(rawJson))

        @JvmStatic
        public fun fromJsonObject(obj: JSONObject): NapaxiConfigSelection {
            val byCapability = obj.optJSONObject("selected_profile_id_by_capability")
            return NapaxiConfigSelection(
                selectedProfileId = obj.optNullableString("selected_profile_id"),
                selectedProfileIdByCapability = byCapability?.keys()?.asSequence()
                    ?.associateWith { key -> byCapability.optString(key) }
                    .orEmpty(),
                systemPrompt = obj.optString("system_prompt"),
                maxToolIterations = obj.optInt("max_tool_iterations", 50),
            )
        }

        @JvmStatic
        public fun fromMap(map: Map<String, *>): NapaxiConfigSelection =
            fromJsonObject(JSONObject(map))
    }
}

public interface NapaxiConfigKeyValueStore {
    public fun read(key: String): String?
    public fun write(key: String, value: String)
    public fun delete(key: String)
}

public interface NapaxiConfigSecretStore {
    public fun read(key: String): String?
    public fun write(key: String, value: String)
    public fun delete(key: String)
}

public class NapaxiConfigStore(
    private val keyValueStore: NapaxiConfigKeyValueStore,
    private val secretStore: NapaxiConfigSecretStore,
) {
    public fun loadProfiles(): List<NapaxiConfigProfile> {
        val raw = keyValueStore.read(PROFILES_KEY)?.takeIf { it.trim().isNotEmpty() } ?: return emptyList()
        return runCatching {
            val array = JSONArray(raw)
            List(array.length()) { index -> array.optJSONObject(index) }
                .mapNotNull { it?.let(NapaxiConfigProfile::fromJsonObject) }
                .filter { it.id.trim().isNotEmpty() }
        }.getOrDefault(emptyList())
    }

    @JvmOverloads
    public fun saveProfile(profile: NapaxiConfigProfile, apiKey: String? = null) {
        val nextProfiles = loadProfiles()
            .filterNot { it.id == profile.id }
            .plus(profile)
        saveProfiles(nextProfiles)
        if (apiKey != null) {
            if (apiKey.isEmpty()) {
                secretStore.delete(apiKeyKey(profile.id))
            } else {
                secretStore.write(apiKeyKey(profile.id), apiKey)
            }
        }
    }

    public fun deleteProfile(profileId: String) {
        saveProfiles(loadProfiles().filterNot { it.id == profileId })
        secretStore.delete(apiKeyKey(profileId))
        val selection = loadSelection()
        val nextByCapability = selection.selectedProfileIdByCapability
            .filterValues { selectedId -> selectedId != profileId }
        saveSelection(
            NapaxiConfigSelection(
                selectedProfileId = selection.selectedProfileId.takeIf { it != profileId },
                selectedProfileIdByCapability = nextByCapability,
                systemPrompt = selection.systemPrompt,
                maxToolIterations = selection.maxToolIterations,
            ),
        )
    }

    public fun loadSelection(): NapaxiConfigSelection {
        val profileIds = loadProfiles().map { it.id }.toSet()
        val raw = keyValueStore.read(SELECTION_KEY)?.takeIf { it.trim().isNotEmpty() }
            ?: return NapaxiConfigSelection()
        val selection = runCatching {
            NapaxiConfigSelection.fromJson(raw)
        }.getOrDefault(NapaxiConfigSelection())
        return normalizeSelection(selection, profileIds)
    }

    public fun saveSelection(selection: NapaxiConfigSelection) {
        val profileIds = loadProfiles().map { it.id }.toSet()
        keyValueStore.write(SELECTION_KEY, normalizeSelection(selection, profileIds).toJson())
    }

    public fun resolveConfig(profileId: String): LlmConfig? {
        val profile = loadProfiles().firstOrNull { it.id == profileId } ?: return null
        return profile.toConfig(apiKey = readApiKey(profileId))
    }

    @JvmOverloads
    public fun resolveSelectedConfig(capabilityId: String? = null): LlmConfig? {
        val selection = loadSelection()
        val selectedProfileId = capabilityId
            ?.let { selection.selectedProfileIdByCapability[it] }
            ?: selection.selectedProfileId
            ?: return null
        return resolveConfig(selectedProfileId)
    }

    public fun readApiKey(profileId: String): String =
        runCatching { secretStore.read(apiKeyKey(profileId)).orEmpty() }.getOrDefault("")

    private fun saveProfiles(profiles: List<NapaxiConfigProfile>) {
        keyValueStore.write(
            PROFILES_KEY,
            JSONArray(profiles.map { it.toJsonObject() }).toString(),
        )
    }

    private fun normalizeSelection(
        selection: NapaxiConfigSelection,
        profileIds: Set<String>,
    ): NapaxiConfigSelection {
        val selectedProfileId = selection.selectedProfileId.takeIf { it != null && profileIds.contains(it) }
        val selectedByCapability = selection.selectedProfileIdByCapability
            .filterValues { profileIds.contains(it) }
        return NapaxiConfigSelection(
            selectedProfileId = selectedProfileId,
            selectedProfileIdByCapability = selectedByCapability,
            systemPrompt = selection.systemPrompt,
            maxToolIterations = selection.maxToolIterations,
        )
    }

    private fun apiKeyKey(profileId: String): String = "$API_KEY_PREFIX$profileId"

    public companion object {
        private const val PROFILES_KEY = "napaxi.config.profiles.v1"
        private const val SELECTION_KEY = "napaxi.config.selection.v1"
        private const val API_KEY_PREFIX = "napaxi.config.api_key."

        @JvmStatic
        public fun memory(): NapaxiConfigStore {
            val store = NapaxiMemoryConfigStore()
            return NapaxiConfigStore(keyValueStore = store, secretStore = store)
        }

        @JvmStatic
        public fun sharedPreferences(context: Context): NapaxiConfigStore =
            NapaxiConfigStore(
                keyValueStore = NapaxiSharedPreferencesConfigStore(context, "napaxi_config"),
                secretStore = NapaxiSharedPreferencesConfigStore(context, "napaxi_config_secrets"),
            )
    }
}

public class NapaxiSharedPreferencesConfigStore(
    context: Context,
    name: String,
) : NapaxiConfigKeyValueStore, NapaxiConfigSecretStore {
    private val preferences = context.applicationContext.getSharedPreferences(name, Context.MODE_PRIVATE)

    override fun read(key: String): String? = preferences.getString(key, null)

    override fun write(key: String, value: String) {
        preferences.edit().putString(key, value).apply()
    }

    override fun delete(key: String) {
        preferences.edit().remove(key).apply()
    }
}

public class NapaxiMemoryConfigStore : NapaxiConfigKeyValueStore, NapaxiConfigSecretStore {
    private val values = LinkedHashMap<String, String>()

    override fun read(key: String): String? = values[key]

    override fun write(key: String, value: String) {
        values[key] = value
    }

    override fun delete(key: String) {
        values.remove(key)
    }
}

private fun JSONObject.toPlainMap(): Map<String, Any?> {
    val values = LinkedHashMap<String, Any?>()
    val keys = keys()
    while (keys.hasNext()) {
        val key = keys.next()
        values[key] = opt(key).toPlainValue()
    }
    return values
}

private fun JSONArray.toPlainList(): List<Any?> =
    List(length()) { index -> opt(index).toPlainValue() }

private fun Any?.toPlainValue(): Any? =
    when (this) {
        null, JSONObject.NULL -> null
        is JSONObject -> this.toPlainMap()
        is JSONArray -> this.toPlainList()
        else -> this
    }

File: packages/flutter/lib/convenience/config_store.dart
import 'dart:convert';

import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../models/config.dart';

/// Persisted LLM profile metadata.
///
/// Sensitive values such as API keys are intentionally excluded from this
/// object and are stored separately by [NapaxiConfigStore].
class NapaxiConfigProfile {
  const NapaxiConfigProfile({
    required this.id,
    required this.name,
    required this.provider,
    required this.model,
    this.baseUrl,
    this.systemPrompt = '',
    this.maxTokens = defaultMaxTokens,
    this.maxToolIterations = 50,
    this.extraHeaders,
    this.userTimezone,
    this.allowedModels,
    this.imageModel,
    this.imageAnalysisModel,
    this.videoModel,
    this.audioModel,
    this.contextEngine = const ContextEngineConfig(),
    this.shellSecurity = const ShellSecurityConfig(),
    this.metadata = const {},
  });

  final String id;
  final String name;
  final String provider;
  final String? baseUrl;
  final String model;
  final String systemPrompt;
  final int maxTokens;
  final int maxToolIterations;
  final String? extraHeaders;
  final String? userTimezone;
  final List<Map<String, String>>? allowedModels;
  final String? imageModel;
  final String? imageAnalysisModel;
  final String? videoModel;
  final String? audioModel;
  final ContextEngineConfig contextEngine;
  final ShellSecurityConfig shellSecurity;
  final Map<String, Object?> metadata;

  Map<String, Object?> toMap() => {
        'id': id,
        'name': name,
        'provider': provider,
        'base_url': baseUrl,
        'model': model,
        'max_tokens': maxTokens,
        'max_tool_iterations': maxToolIterations,
        'extra_headers': extraHeaders,
        if (userTimezone != null && userTimezone!.trim().isNotEmpty)
          'user_timezone': userTimezone,
        if (systemPrompt.trim().isNotEmpty) 'system_prompt': systemPrompt,
        if (allowedModels != null) 'allowed_models': allowedModels,
        if (imageModel != null) 'image_model': imageModel,
        if (imageAnalysisModel != null)
          'image_analysis_model': imageAnalysisModel,
        if (videoModel != null) 'video_model': videoModel,
        if (audioModel != null) 'audio_model': audioModel,
        'context_engine': contextEngine.toMap(),
        'shell_security': shellSecurity.toMap(),
        if (metadata.isNotEmpty) 'metadata': metadata,
      };

  factory NapaxiConfigProfile.fromMap(Map<String, Object?> map) {
    return NapaxiConfigProfile(
      id: map['id'] as String? ?? '',
      name: map['name'] as String? ?? '',
      provider: map['provider'] as String? ?? '',
      baseUrl: map['base_url'] as String?,
      model: map['model'] as String? ?? '',
      systemPrompt: map['system_prompt'] as String? ?? '',
      maxTokens: map['max_tokens'] as int? ?? defaultMaxTokens,
      maxToolIterations: map['max_tool_iterations'] as int? ?? 50,
      extraHeaders: map['extra_headers'] as String?,
      userTimezone: map['user_timezone'] as String? ??
          map['userTimeZone'] as String? ??
          map['timeZoneId'] as String?,
      allowedModels: (map['allowed_models'] as List?)
          ?.map((e) => Map<String, String>.from(e as Map))
          .toList(),
      imageModel: map['image_model'] as String?,
      imageAnalysisModel: map['image_analysis_model'] as String?,
      videoModel: map['video_model'] as String?,
      audioModel: map['audio_model'] as String?,
      contextEngine: map['context_engine'] is Map
          ? ContextEngineConfig.fromMap(
              Map<String, dynamic>.from(map['context_engine'] as Map),
            )
          : const ContextEngineConfig(),
      shellSecurity: map['shell_security'] is Map
          ? ShellSecurityConfig.fromMap(
              Map<String, dynamic>.from(map['shell_security'] as Map),
            )
          : const ShellSecurityConfig(),
      metadata: Map<String, Object?>.from(
        map['metadata'] as Map? ?? const {},
      ),
    );
  }

  LlmConfig toConfig({required String apiKey}) {
    return LlmConfig(
      provider: provider,
      apiKey: apiKey,
      baseUrl: baseUrl,
      model: model,
      systemPrompt: systemPrompt,
      maxTokens: maxTokens,
      maxToolIterations: maxToolIterations,
      extraHeaders: extraHeaders,
      userTimezone: userTimezone,
      allowedModels: allowedModels,
      imageModel: imageModel,
      imageAnalysisModel: imageAnalysisModel,
      videoModel: videoModel,
      audioModel: audioModel,
      contextEngine: contextEngine,
      shellSecurity: shellSecurity,
    );
  }
}

/// The user's active configuration selection: chosen profile(s), system
/// prompt, and tool-iteration limit.
class NapaxiConfigSelection {
  const NapaxiConfigSelection({
    this.selectedProfileId,
    this.selectedProfileIdByCapability = const {},
    this.systemPrompt = '',
    this.maxToolIterations = 50,
  });

  final String? selectedProfileId;
  final Map<String, String> selectedProfileIdByCapability;
  final String systemPrompt;
  final int maxToolIterations;

  Map<String, Object?> toMap() => {
        'selected_profile_id': selectedProfileId,
        'selected_profile_id_by_capability': selectedProfileIdByCapability,
        'system_prompt': systemPrompt,
        'max_tool_iterations': maxToolIterations,
      };

  factory NapaxiConfigSelection.fromMap(Map<String, Object?> map) {
    return NapaxiConfigSelection(
      selectedProfileId: map['selected_profile_id'] as String?,
      selectedProfileIdByCapability: Map<String, String>.from(
        map['selected_profile_id_by_capability'] as Map? ?? const {},
      ),
      systemPrompt: map['system_prompt'] as String? ?? '',
      maxToolIterations: map['max_tool_iterations'] as int? ?? 50,
    );
  }
}

/// Storage seam for non-secret config values (profiles, selection).
abstract class NapaxiConfigKeyValueStore {
  Future<String?> read(String key);

  Future<void> write(String key, String value);

  Future<void> delete(String key);
}

/// Secure-storage seam for sensitive config values such as API keys.
abstract class NapaxiConfigSecretStore {
  Future<String?> read(String key);

  Future<void> write(String key, String value);

  Future<void> delete(String key);
}

/// Persists LLM config profiles and the active selection, splitting non-secret
/// values and secrets (API keys) across separate backing stores.
class NapaxiConfigStore {
  NapaxiConfigStore({
    required NapaxiConfigKeyValueStore keyValueStore,
    required NapaxiConfigSecretStore secretStore,
  })  : _keyValueStore = keyValueStore,
        _secretStore = secretStore;

  NapaxiConfigStore.memory()
      : this(
          keyValueStore: _MemoryConfigStore(),
          secretStore: _MemoryConfigStore(),
        );

  static final NapaxiConfigStore instance = NapaxiConfigStore(
    keyValueStore: _SharedPreferencesConfigStore(),
    secretStore: _FlutterSecureConfigStore(),
  );

  static const _profilesKey = 'napaxi.config.profiles.v1';
  static const _selectionKey = 'napaxi.config.selection.v1';
  static const _apiKeyPrefix = 'napaxi.config.api_key.';

  final NapaxiConfigKeyValueStore _keyValueStore;
  final NapaxiConfigSecretStore _secretStore;

  Future<List<NapaxiConfigProfile>> loadProfiles() async {
    final raw = await _keyValueStore.read(_profilesKey);
    if (raw == null || raw.trim().isEmpty) return const [];
    final decoded = jsonDecode(raw);
    if (decoded is! List) return const [];
    return decoded
        .whereType<Map>()
        .map((item) => NapaxiConfigProfile.fromMap(
              Map<String, Object?>.from(item),
            ))
        .where((profile) => profile.id.trim().isNotEmpty)
        .toList();
  }

  Future<void> saveProfile(
    NapaxiConfigProfile profile, {
    String? apiKey,
  }) async {
    final profiles = await loadProfiles();
    final nextProfiles = [
      for (final item in profiles)
        if (item.id != profile.id) item,
      profile,
    ];
    await _saveProfiles(nextProfiles);
    if (apiKey != null) {
      if (apiKey.isEmpty) {
        await _secretStore.delete(_apiKeyKey(profile.id));
      } else {
        await _secretStore.write(_apiKeyKey(profile.id), apiKey);
      }
    }
  }

  Future<void> deleteProfile(String profileId) async {
    final profiles = await loadProfiles();
    await _saveProfiles(
      profiles.where((profile) => profile.id != profileId).toList(),
    );
    await _secretStore.delete(_apiKeyKey(profileId));
    final selection = await loadSelection();
    final nextCapabilitySelection =
        Map<String, String>.from(selection.selectedProfileIdByCapability)
          ..removeWhere((_, selectedId) => selectedId == profileId);
    await saveSelection(NapaxiConfigSelection(
      selectedProfileId: selection.selectedProfileId == profileId
          ? null
          : selection.selectedProfileId,
      selectedProfileIdByCapability: nextCapabilitySelection,
      systemPrompt: selection.systemPrompt,
      maxToolIterations: selection.maxToolIterations,
    ));
  }

  Future<NapaxiConfigSelection> loadSelection() async {
    final profiles = await loadProfiles();
    final profileIds = profiles.map((profile) => profile.id).toSet();
    final raw = await _keyValueStore.read(_selectionKey);
    if (raw == null || raw.trim().isEmpty) {
      return const NapaxiConfigSelection();
    }
    final decoded = jsonDecode(raw);
    if (decoded is! Map) return const NapaxiConfigSelection();
    final selection = NapaxiConfigSelection.fromMap(
      Map<String, Object?>.from(decoded),
    );
    return _normalizeSelection(selection, profileIds);
  }

  Future<void> saveSelection(NapaxiConfigSelection selection) async {
    final profiles = await loadProfiles();
    final profileIds = profiles.map((profile) => profile.id).toSet();
    final normalized = _normalizeSelection(selection, profileIds);
    await _keyValueStore.write(_selectionKey, jsonEncode(normalized.toMap()));
  }

  Future<LlmConfig?> resolveConfig(String profileId) async {
    final profiles = await loadProfiles();
    NapaxiConfigProfile? selectedProfile;
    for (final profile in profiles) {
      if (profile.id == profileId) {
        selectedProfile = profile;
        break;
      }
    }
    if (selectedProfile == null) return null;
    String apiKey;
    try {
      apiKey = await _secretStore.read(_apiKeyKey(profileId)) ?? '';
    } catch (_) {
      apiKey = '';
    }
    return selectedProfile.toConfig(apiKey: apiKey);
  }

  Future<String> readApiKey(String profileId) async {
    try {
      return await _secretStore.read(_apiKeyKey(profileId)) ?? '';
    } catch (_) {
      return '';
    }
  }

  Future<void> _saveProfiles(List<NapaxiConfigProfile> profiles) {
    return _keyValueStore.write(
      _profilesKey,
      jsonEncode(profiles.map((profile) => profile.toMap()).toList()),
    );
  }

  NapaxiConfigSelection _normalizeSelection(
    NapaxiConfigSelection selection,
    Set<String> profileIds,
  ) {
    final selectedProfileId = profileIds.contains(selection.selectedProfileId)
        ? selection.selectedProfileId
        : null;
    final capabilitySelection =
        Map<String, String>.from(selection.selectedProfileIdByCapability)
          ..removeWhere((_, profileId) => !profileIds.contains(profileId));
    return NapaxiConfigSelection(
      selectedProfileId: selectedProfileId,
      selectedProfileIdByCapability: capabilitySelection,
      systemPrompt: selection.systemPrompt,
      maxToolIterations: selection.maxToolIterations,
    );
  }

  String _apiKeyKey(String profileId) => '$_apiKeyPrefix$profileId';
}

class _SharedPreferencesConfigStore implements NapaxiConfigKeyValueStore {
  @override
  Future<String?> read(String key) async {
    final preferences = await SharedPreferences.getInstance();
    return preferences.getString(key);
  }

  @override
  Future<void> write(String key, String value) async {
    final preferences = await SharedPreferences.getInstance();
    await preferences.setString(key, value);
  }

  @override
  Future<void> delete(String key) async {
    final preferences = await SharedPreferences.getInstance();
    await preferences.remove(key);
  }
}

class _FlutterSecureConfigStore implements NapaxiConfigSecretStore {
  static const _storage = FlutterSecureStorage();

  @override
  Future<String?> read(String key) => _storage.read(key: key);

  @override
  Future<void> write(String key, String value) {
    return _storage.write(key: key, value: value);
  }

  @override
  Future<void> delete(String key) => _storage.delete(key: key);
}

class _MemoryConfigStore
    implements NapaxiConfigKeyValueStore, NapaxiConfigSecretStore {
  final Map<String, String> _values = {};

  @override
  Future<String?> read(String key) async => _values[key];

  @override
  Future<void> write(String key, String value) async {
    _values[key] = value;
  }

  @override
  Future<void> delete(String key) async {
    _values.remove(key);
  }
}
